halo2_proofs/src/dev/cost_model.rs

```diff
@@ -0,0 +1,633 @@
+            // value(X) = value(X) * y + (1 - (l_last(X) + l_blind(X))) * (
+            //   z_i(\omega X) \prod_j (p(X) + \beta s_j(X) + \gamma)
+            // - z_i(X) \prod_j (p(X) + \delta^j \beta X + \gamma)
+            // ).
+            tmp_num_add_lag += 2;
+            tmp_num_mul_lag += 2;
+
+            num_add_lag += tmp_num_add_lag * num_perm_slices;
+            num_mul_lag += tmp_num_mul_lag * num_perm_slices;
+
+            // beta_term *= &extended_omega;
+            num_mul_lag += 1;
+        }
+
+        // Lookups
+        let num_lookups = pk.get_vk().get_cs().lookups.len();
+        // a_minus_s
+        num_add_lag += num_lookups;
+        // value(X) = value(X) * y + l_0(X) * (1 - z(X))
+        num_add_lag += 2 * num_lookups;
+        num_mul_lag += 2 * num_lookups;
+        // value(X) = value(X) * y + l_last(X) * (z(X)^2 - z(X))
+        num_add_lag += 2 * num_lookups;
+        num_mul_lag += 3 * num_lookups;
+        // value(X) = value(X) * y + (1 - (l_last(X) + l_blind(X))) * (
+        //   z(\omega X) (a'(X) + \beta) (s'(X) + \gamma)
+        //   - z(X) (\theta^{m-1} a_0(X) + ... + a_{m-1}(X) + \beta)
+        //          (\theta^{m-1} s_0(X) + ... + s_{m-1}(X) + \gamma)
+        // )
+        num_add_lag += 4 * num_lookups;
+        num_mul_lag += 5 * num_lookups;
+        // value(X) = value(X) * y + l_0(X) * (a'(X) - s'(X))
+        num_add_lag += 1 * num_lookups;
+        num_mul_lag += 2 * num_lookups;
+        // (1 - (l_last + l_blind)) * (a′(X) − s′(X))⋅(a′(X) − a′(\omega^{-1} X)) = 0
+        num_add_lag += 2 * num_lookups;
+        num_mul_lag += 3 * num_lookups;
+
+        num_add_lag *= l;
+        num_mul_lag *= l;
+
+        (num_add_lag, num_mul_lag)
+    }
+}
+
+/// estimate is to estimate the prover time, peek memory usage and aggregate circuit size.
+pub fn estimate<E: Engine, ConcreteCircuit: Circuit<E::Scalar>>(
+    circuit: ConcreteCircuit,
+    k: usize,
+) -> EstimateResult {
+    // Generate small vk & pk
+    let params: Params<E::G1Affine> = Params::<E::G1Affine>::unsafe_setup::<E>(15_u32);
+    let vk = keygen_vk(&params, &circuit).expect("keygen_vk should not fail");
+    let pk = keygen_pk(&params, vk, &circuit).expect("keygen_pk should not fail");
+
+    let l = 1;
+
+    // Initialize the polynomial commitment parameters
+    let cs = pk.get_vk().get_cs();
+
+    let generate_fake_params = |k| {
+        let s = E::Scalar::random(OsRng);
+        let rand_c1 = <E::G1Affine as PrimeCurveAffine>::generator() * s;
+        let rand_c2 = <E::G2Affine as PrimeCurveAffine>::generator() * s;
+        let rand_c1: E::G1Affine = rand_c1.into();
+        let n = 1 << k;
+        Params {
+            k: k as u32,
+            n: n as u64,
+            g: (0..n).map(|_| rand_c1).collect(),
+            g_lagrange: (0..n).map(|_| rand_c1).collect(),
+            additional_data: Vec::from(rand_c2.to_bytes().as_ref()),
+        }
+    };
+
+    let params = generate_fake_params(k);
```

Comment:
By applying the above comment, this fake_params is not needed anymore.

Reply:
The previous `params` is the one with fixed smaller `k`, while `fake_params` is to generate one with larger `k`.